src/CFamily/Rules/RulesConfigFixup.cs

```diff
@@ -165,34 +198,54 @@ public RulesSettings Apply(RulesSettings input)
                 * user disables the corresponding "new" rule S787.
             In that case, we'll warn in the output window that the legacy setting is being ignored.
 
-             */
+            */
             
             var modifiedSettings = new RulesSettings
             {
                 Rules = new Dictionary<string, RuleConfig>(input.Rules, input.Rules.Comparer)
             };
 
-            foreach (var inputKey in modifiedSettings.Rules.Keys.ToArray())
+            TranslateLegacyRuleKeys(modifiedSettings);
+            DisableExcludedRules(modifiedSettings);
+
+            return modifiedSettings;
+        }
+
+        private void TranslateLegacyRuleKeys(RulesSettings settings)
+        {
+            foreach (var inputKey in settings.Rules.Keys.ToArray())
             {
                 if (fullLegacyToNewKeyMap.TryGetValue(inputKey, out var newKey))
                 {
-                    var inputConfig = modifiedSettings.Rules[inputKey];
-                    modifiedSettings.Rules.Remove(inputKey);
+                    var inputConfig = settings.Rules[inputKey];
+                    settings.Rules.Remove(inputKey);
 
                     // There might already be a setting with the new key. If so, we'll keep it and drop the legacy key setting.
-                    if (modifiedSettings.Rules.ContainsKey(newKey))
+                    if (settings.Rules.ContainsKey(newKey))
                     {
                         logger.WriteLine(Resources.CFamily_DuplicateLegacyAndNewRuleKey, inputKey, newKey);
                     }
                     else
                     {
                         logger.LogDebug($"[CFamily] Translating legacy rule key: {inputKey} -> {newKey}");
-                        modifiedSettings.Rules[newKey] = inputConfig;
+                        settings.Rules[newKey] = inputConfig;
                     }
                 }
             }
+        }
 
-            return modifiedSettings;
+        /// <summary>
+        /// Marks all excluded rules as disabled, adding them to the settings if necessary
```

Comment:
Moved from `DynamicCFamilyRulesConfig`.
It's also slightly simpler, since the code that creates a shallow copy of the settings is done above.